ops/model.py

```diff
@@ -797,6 +961,313 @@ def __init__(self, name: str, address_info: '_AddressDict'):
+                    'secret secret ID {!r} should start with `secret:`'.format(id)
+                )
+            if id.strip() != id:
+                # FIXME one of the hook tools returns some whitespace around
+                #  the secret-id. We should catch that earlier and trim it there.
+                logger.warning(
+                    'invalid characters in secret ID '
+                    '{!r}: something wrong'.format(id)
+                )
+                id = id.strip()
+
+        self._id = id
+        self._label = label
+        self._revision = revision
+        self._am_owner = am_owner
+
+    @property
+    def id(self) -> Optional[str]:
+        """Unique identifier for this secret.
+
+        The secret ID will be known (not None) only if you either:
+            - obtain this secret by `self.[unit|app].add_secret(...)`
+            - obtain this secret by `self.model.get_secret(id=<some-id>)`
+            - obtain this secret from a SecretEvent instance.
+
+        If you obtained this secret by `self.model.get_secret(label=<some-label>)`,
+            then the secret ID will be unknown.
+            The secret will then be locally addressable only by its label.
+        """
+        return self._id
+
+    @property
+    def label(self) -> Optional[str]:
+        """Label by which this secret is locally known.
+
+        The label will be not None only if you either:
+            - obtain this secret by `self.[unit|app].add_secret(label=<some-label>, ...)`
+            - obtain this secret by `self.model.get_secret(label=<some-label>)`
+            - obtain this secret from a SecretEvent instance.
+
+        If you obtained this secret by `self.model.get_secret(id=<some-id>)`,
+            then the label for this secret will be unknown.
+            The secret will then be locally addressable only by its ID.
+
+        This label is locally unique.
+        """
+        return self._label
+
+    @property
+    def revision(self) -> int:
+        """The revision number for this secret."""
+        if self._revision is None:
+            if self._am_owner:
+                # use _revision if you know what you're doing (testing, presumably)
+                raise RuntimeError("This secret has no revision.")
+            else:
+                raise RuntimeError("Only secret owners can see the revision number.")
+        return self._revision
+
+    def __str__(self):
+        if self.label:
+            return '<{}({})>'.format(self.id, self.label)
+        return '<{}>'.format(self.id)
+
+    @staticmethod
+    def _validate_content(content: Optional[Dict[str, str]]):
+        if not content:
+            # fixme: ATM you can't secret-set without content (e.g. to only update the label),
+            #  but that is a bug in juju.
+            #  when that is fixed, `secret-set [id] --label X` will be valid.
+            raise ModelError('missing secret value or filename')
+        for key, _ in content.items():
```

Comment:
We could just use `content.keys()` instead of throwing away the second value.

And not trying to be too "golfy", but:

```python
for [k in content.keys() if not _Secret.secret_key_re.match(k)]:
  raise ModelError(...
  ```
  
Feels a little more fluent to me. Your call. I just (over)use comprehensions

Reply:
Good point. Not quite as golfy, but it allows to catch all bad keys at once:
```
invalid_secret_keys = filter(lambda key: not _Secret.secret_key_re.match(key),
                                     content.keys())
        if invalid_secret_keys:
            raise ModelError(
```